### DISCUSSION TITLE
WebGPU object selection and picking

### DISCUSSION THREAD

**Discussion by danchitnis** [gpuweb/gpuweb#2654] [Q&A] [answered]
Hello, I was wondering if WebGPU has any native methods for interactive selection (e.g. user mouse input) of drawn objects on the canvas, basically similar to OpenGL ["picking"](https://www.glprogramming.com/red/chapter13.html) and the related WebGL color based hacks. Thanks.

**Comment by ghadeeras** [comment]
Hello. This is not an answer to your question, as I am not one of the contributors to the WebGPU specs, and I myself am interested in the appropriate answer to your question, but my guess is that one way to achieve that is by performing, upon user input, a render pass onto an offscreen texture with "coded" colors using a separate shader, then reading the texture color at, say, the mouse's location.

This was how I did it in my toyish sculpting tool: https://github.com/ghadeeras/ghadeeras.github.io/blob/master/src/scalar-field/picker.gpu.ts ... Mind you, this is not a good example code, as I am using a bunch experimental self-made libraries that could render things a bit obscure. The shader for this is: https://github.com/ghadeeras/ghadeeras.github.io/blob/master/shaders/picker.wgsl

I am curious to see if there are better/more-idiomatic ways of doing this in WebGPU.

**Reply by danchitnis** [reply]
Thanks @ghadeeras for your suggestion. I believe the method that you described has been used in WebGL as a hack for object selection. I guess the main problem with that is the selection of the objects which are occluded from the view (same view, different depths). I would be inserted to know if there is a more native method of doing this in WebGPU.

**Comment by kdashg** [comment, marked as answer]
There is no specific picking API in WebGPU. Like in WebGL, this is something you must do yourself!
